### ISSUE TITLE
Configure delay for 'onRouteUpdateDelayed' hook ?

### ISSUE THREAD

**Issue by ifahrentholz** [gatsbyjs/gatsby#23755]
## Summary
Is it possible to configure the delay for 'onRouteUpdateDelayed' hook ?

## Relevant information

Documentation link: https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdateDelayed

For my page I implement a loading indicator. Therefore I would like to configure the default delay for the 'onRouteUpdateDelayed' hook. Currently I use it inside the gatsby-browser.js like so:

```
exports.onRouteUpdateDelayed = () => {
  window.___emitter.emit('onRouteUpdateDelayed');
};
```

The default delay is 1s but i would like to show the spinner earlier like after 500ms or so..

## Gatsby info

```
  System:
    OS: macOS 10.15.4
    CPU: (8) x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
    Shell: Unknown
  Binaries:
    Node: 12.13.0 - ~/.nvm/versions/node/v12.13.0/bin/node
    npm: 6.13.7 - ~/.nvm/versions/node/v12.13.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 81.0.4044.129
    Firefox: 75.0
    Safari: 13.1
  npmPackages:
    gatsby: ^2.20.12 => 2.20.15 
    gatsby-image: ^2.3.1 => 2.3.1 
    gatsby-plugin-layout: ^1.2.2 => 1.2.2 
    gatsby-plugin-manifest: ^2.3.3 => 2.3.3 
    gatsby-plugin-offline: ^3.1.2 => 3.1.2 
    gatsby-plugin-react-helmet: ^3.2.1 => 3.2.1 
    gatsby-plugin-react-svg: 3.0.0 => 3.0.0 
    gatsby-plugin-sass: ^2.2.1 => 2.2.1 
    gatsby-plugin-sharp: ^2.5.3 => 2.5.3 
    gatsby-source-filesystem: 2.3.0 => 2.3.0 
    gatsby-source-multi-api: 1.0.3 => 1.0.3 
    gatsby-transformer-json: 2.4.0 => 2.4.0 
    gatsby-transformer-sharp: ^2.4.3 => 2.4.3 
```

### File contents (if changed)

`gatsby-config.js`: N/A 
`package.json`: N/A 
`gatsby-node.js`: N/A 
`gatsby-browser.js`: 
```
exports.onRouteUpdateDelayed = () => {
  window.___emitter.emit('onRouteUpdateDelayed');
};

exports.onRouteUpdate = () => {
  window.___emitter.emit('onRouteUpdate');
};
```
`gatsby-ssr.js`: N/A

**Comment by mjsarfatti** [comment]
I have the same need. I feel like 1s is a _long_ time for a website that's supposed to be instant.

Personally, I'd like to show a loading indicator after 250ms. Some smart folks at Google even consider any delay above 100ms as deserving an indicator (and they developed the whole gmail experience around this).

I don't have a problem with it being 1s by default, but I feel it should be customizable.

**Comment by jzabala** [comment]
Hi, I would like to give this one a try but, I where would the timeout option be, an env, in the `gatsby-config`?

**Comment by ehrencrona** [comment]
I looked at this briefly. I don't think there is any sensible place to put that configuration. Also, you might conceivably want different intervals on different pages so I don't think it should be a global configuration.

I think the cleanest solution would be to add a new API `onRouteUpdateStart` that is triggered before the page load when navigating. The project code can then determine  determine if loading is slow by looking for an `onRouteUpdateStart` that has not been followed by an `onRouteUpdate` in whatever time period they prefer. 

(It's also worth considering what happens if the page fails to load; looking at `packages/gatsby/cache-dir/navigation.js` the error case seems unhandled at the moment; maybe an `onRouteUpdateError` would be in order?)

**Comment by jzabala** [comment]
> I looked at this briefly. I don't think there is any sensible place to put that configuration. Also, you might conceivably want different intervals on different pages so I don't think it should be a global configuration.
> 
> I think the cleanest solution would be to add a new API `onRouteUpdateStart` that is triggered before the page load when navigating. The project code can then determine determine if loading is slow by looking for an `onRouteUpdateStart` that has not been followed by an `onRouteUpdate` in whatever time period they prefer.

Hi @ehrencrona, thanks to your comments I now see that in reality we don't need to do anything here.

The new api that you proposed `onRouteUpdateStart` already exists with the name [`onPreRouteUpdate`](https://www.gatsbyjs.org/docs/browser-apis/#onPreRouteUpdate). So, people wanting to bail out of the 1s delay can just use a combination of `onPreRouteUpdate` and `onRouteUpdate`.

> (It's also worth considering what happens if the page fails to load; looking at `packages/gatsby/cache-dir/navigation.js` the error case seems unhandled at the moment; maybe an `onRouteUpdateError` would be in order?)

Yeah, your are right. Maybe we should open a new issue for that and close this one as answered (if the answer about using the available apis is good enough of course)?

**Comment by mjsarfatti** [comment]
I think 99% of the times the interval would be the same on all pages. Is it really not possible to place the config in `gatsby-config.js`?

I believe having this "baked in" gatsby's navigation can avoid many problems to users who would have to reinvent the implementation every time.

**Comment by anasbouzid** [comment]
> The new api that you proposed `onRouteUpdateStart` already exists with the name [`onPreRouteUpdate`](https://www.gatsbyjs.org/docs/browser-apis/#onPreRouteUpdate). So, people wanting to bail out of the 1s delay can just use a combination of `onPreRouteUpdate` and `onRouteUpdate`.

@jzabala I don't think `onPreRouteUpdate` is the correct API. I've done some testing on https://xzvto-8000.sse.codesandbox.io/page-2/ ([source](https://codesandbox.io/s/jolly-newton-xzvto?file=/gatsby-browser.js)) and on a simulated slow network, the `onRouteUpdateDelayed` event is fired before `onPreRouteUpdate` (make sure you click a link before it's prefetched)

**Comment by gethinwebster** [comment]
Suggest reopening this as I've also come across this problem - I want to be able to adjust this delay, and the suggestion of using `onPreRouteUpdate` is incorrect, as this hook gets called too late (after the new page is loaded, but before it's rendered)

**Comment by jedevc** [comment]
Would it be possible to re-open this issue? I don't think the underlying problem is resolved:

- As mentioned above, `onPreRouteUpdate` is actually triggered when the page has loaded, which means we can't hook it to start configuring a delay (and this probably can't change because of backwards-compatibility)
- There's no `onRouteUpdateStart` which allows for hooking before the page load starts (something like that could be added around https://github.com/gatsbyjs/gatsby/blob/5c22948094d98f3d4ea90253c80fc78d9f53cc20/packages/gatsby/cache-dir/navigation.js#L78)
- And there's no option to configure the 1000ms delay through gatsby config or API.

This is kind of bugging me, so I'd be happy to take a look at a patch - I think *ideally* from a user's perspective both an `onRouteUpdateStart` and delay configuration option could be provided, but would probably be worth picking a single one to start with.

**Comment by sheharyarn** [comment]
Bumping this as I would also like a hook similar to `onRouteUpdateStart`.

**Comment by helmax-y** [comment]
Opportunity to configure delay is still missing. I want to implement screen blur on page transition which should act like loader until page is loading. Adding such blur only after 1000ms doesn't work good, it should be shown immediately after link is clicked.

Please contribute into this improvement once you have capacity. `onRouteUpdateStart` sound like a good option
